### PR TITLE
[GraphBolt] Add sampled sub graph

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add sampled sub graph, which should be the common result graph for sampling used in graphbolt.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+class SampledSubGraph:
+    r"""Class for sampled subgraph."""
+
+    def __init__(
+        self,
+        node_pairs,
+        reverse_column_node_ids=None,
+        reverse_row_node_ids=None,
+        reverse_edge_ids=None,
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:12]: Add type.

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    ):
+        """ ""Initialize the SampledSubGraph object. In the context of a
+        heterogeneous graph, each field should be of `Dict` type. For
+        homogeneous graphs, each field should correspond to its respective
+        value type."
+
+        Parameters
+        ----------
+        node_pairs :  Tuple[Tensor] or Dict[(str, str, str), Tuple[Tensor]]
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:23]: Do we prefer COO here or CSC?

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            Row's reverse node ids in the original graph. A graph structure
+            can be treated as a coordinated row and column pair, and this is
+            the mapped ids of the row.
+            - If `reverse_row_node_ids` is a tensor: It represents the
+                original node ids.
+            - If `reverse_row_node_ids` is a dictionary: The keys should be
+                node type and the values should be corresponding original
+                heterogeneous node ids.
+        reverse_edge_ids: Optional[Tensor or Dict[(str, str, str), Tensor]]
+            Reverse edge ids in the original graph.This is useful when edge
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:49]:
```suggestion
            Reverse edge ids in the original graph. This is useful when edge
```

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            features are needed.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:50]: Clarify when it will be present and when it will be absent.

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
+"""Heterogeneous sampled subgraph."""
+# pylint: disable= invalid-name
+from collections.abc import Mapping
+
+
+class SampledSubGraph:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:6]: Instead of making it a class, let's make it a API first, and then implementation.

Comment (frozenbugs) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:6]: And move it to graphbolt/ top level folder.

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 1 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:14]: added

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    r"""Class for sampled subgraph."""
+
+    def __init__(
+        self,
+        node_pairs,
+        reverse_column_node_ids=None,
+        reverse_row_node_ids=None,
+        reverse_edge_ids=None,
+    ):
+        """ ""Initialize the SampledSubGraph object. In the context of a
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:16]: fixed

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        heterogeneous graph, each field should be of `Dict` type. For
+        homogeneous graphs, each field should correspond to its respective
+        value type."
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:19]: fixed

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        Parameters
+        ----------
+        node_pairs :  Tuple[Tensor] or Dict[(str, str, str), Tuple[Tensor]]
+            Node pairs representing source-destination edges.
+            - If `node_pairs` is a tuple: It should be in the format ('u', 'v')
+                representing source and destination pairs.
+            - If `node_pairs` is a dictionary: The keys should be edge type and
+                the values should be corresponding node pairs. The ids inside
+                is heterogeneous ids.
+        reverse_column_node_ids : Optional[Tensor or Dict[str, Tensor]]
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:30]: Yes, this is optional. Because some algorithm will not compact columns, we can't assume it is compressed. 
If the sampled graph is of format `CSC`,  it is the input seed nodes.

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            Column's reverse node ids in the original graph. A graph structure
+            can be treated as a coordinated row and column pair, and this is
+            the mapped ids of the column.
+            - If `reverse_column_node_ids` is a tensor: It represents the
+                original node ids.
+            - If `reverse_column_node_ids` is a dictionary: The keys should be
+                node type and the values should be corresponding original
+                heterogeneous node ids.
+        reverse_row_node_ids : Optional[Tensor or Dict[str, Tensor]]
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:39]: added

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 2 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:23]: Both can be supported in this structure, because both `COO` and  `CSC` are 2 tensors. In our canonical case, the subgraph isn't compacted. Therefore, we opt for `COO` over `CSC` to avoid numerous gaps that would occur with `CSC`.

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 3 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:49]: added

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 4 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:50]: fixed

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 5 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py:6]: I'm not quite understand how to make it as an API, because this struct is needed as the result of sampling.

File: python/dgl/graphbolt/__init__.py [hunk 10]
```diff
@@ -14,7 +14,7 @@
 from .dataset import *
 from .dataloader import *
 from .subgraph_sampler import *
-
+from .sampled_sub_graph import *
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/__init__.py:17]: Better to use `.sampled_subgraph`?

File: python/dgl/graphbolt/sampled_sub_graph.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+class SampledSubGraph:
+    r"""Class for sampled subgraph."""
+
+    def __init__(
+        self,
+        node_pairs: Union[
+            Dict[Tuple[str, str, str], Tuple[torch.tensor, torch.tensor]],
+            Tuple[torch.tensor, torch.tensor],
+        ],
+        reverse_column_node_ids: Union[
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/sampled_sub_graph.py:18]: I feel the word `reverse` is not straightforward. Can we use `original`?

File: python/dgl/graphbolt/sampled_sub_graph.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            Dict[str, torch.tensor], torch.tensor
+        ] = None,
+        reverse_row_node_ids: Union[
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/sampled_sub_graph.py:21]: Since we are not in the scope of a `CSC` graph, it is strange to use `row` and `column`. Can we use `src` and `dst`?

File: python/dgl/graphbolt/sampled_sub_graph.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+            row IDs match these compacted ones.
+        reverse_edge_ids: Optional[Tensor or Dict[(str, str, str), Tensor]]
+            Reverse edge ids in the original graph. This is useful when edge
+            features are needed.
+            - If `reverse_edge_ids` is a tensor: It represents the
+                original edge ids.
+            - If `reverse_edge_ids` is a dictionary: The keys should be
+                edge type and the values should be corresponding original
+                heterogeneous edge ids.
+
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/sampled_sub_graph.py:72]: Better to give an example here.

File: python/dgl/graphbolt/sampled_sub_graph.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
+"""Heterogeneous sampled subgraph."""
+# pylint: disable= invalid-name
+from collections.abc import Mapping
+from typing import Dict, Tuple, Union
+
+import torch
+
+
+class SampledSubGraph:
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/sampled_sub_graph.py:9]: Since this class is purely for storing data, can we use `@dataclass` or `NamedTuple`?

Comment (BarclayII) [reply python/dgl/graphbolt/sampled_sub_graph.py:9]: Also change `SubGraph` to `Subgraph`.

File: python/dgl/graphbolt/sampled_sub_graph.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+    r"""Class for sampled subgraph."""
+
+    def __init__(
+        self,
+        node_pairs: Union[
+            Dict[Tuple[str, str, str], Tuple[torch.tensor, torch.tensor]],
```
Comment (BarclayII) [inline python/dgl/graphbolt/sampled_sub_graph.py:15]: The typing annotation should be `torch.Tensor` rather than `torch.tensor`, which is a function.

File: python/dgl/graphbolt/sampled_sub_graph.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+    ):
+        """Initialize the SampledSubGraph object. In the context of a
+        heterogeneous graph, each field should be of `Dict` type. For
+        homogeneous graphs, each field should correspond to its respective
+        value type.
+
+        Parameters
+        ----------
+        node_pairs :  Tuple[Tensor, Tensor] or Dict[(str, str, str), Tuple[Tensor, Tensor]]
+            Node pairs representing source-destination edges.
```
Comment (BarclayII) [inline python/dgl/graphbolt/sampled_sub_graph.py:36]: Would `SampleSubgraph` support subgraphs of a multigraph?  If so, how do we distinguish multi-edges for a subgraph from a multigraph if we only use node pairs?  If not, are we going to make a new class to support it?

**Review by BarclayII — CHANGES_REQUESTED** [review]
I'm concerned about using node pairs to represent edges.  If we only consider simple graphs, we should state so in this class.  Even so, I wonder what we are going to do for subgraphs of multi-graphs.

File: python/dgl/graphbolt/sampled_sub_graph.py (hunk 16 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/sampled_sub_graph.py:36]: Yes, we support multigraph as well, there is a optional "reverse_edge_ids" that can be used to distinguish multiple edges.

File: python/dgl/graphbolt/__init__.py (hunk 10 above)
Comment (peizhou001) [reply python/dgl/graphbolt/__init__.py:17]: make sense, fixed

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 5 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py]: See this https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/dataset.py as an example.

File: python/dgl/graphbolt/sampled_sub_graph.py (hunk 14 above)
Comment (peizhou001) [reply python/dgl/graphbolt/sampled_sub_graph.py:9]: fixed

File: python/dgl/graphbolt/sampled_sub_graph.py (hunk 15 above)
Comment (peizhou001) [reply python/dgl/graphbolt/sampled_sub_graph.py:15]: fixed

File: python/dgl/graphbolt/sampled_sub_graph.py (hunk 12 above)
Comment (peizhou001) [reply python/dgl/graphbolt/sampled_sub_graph.py:21]: Actually we don't want to limit the format of the sub graph, it could be a `CSC`, `CSR` or `COO`, and all of them can be represented as 2 tensors. From this perspective,  it's better to use `row` and `column`?

**Review by frozenbugs — COMMENTED** [review]
Can you separate it to 2 PRs? 1 for refactoring, the other for adding API.

File: python/dgl/graphbolt/graph_storage/sampled_sub_graph.py (hunk 5 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/sampled_sub_graph.py]: Changed to API style.

File: python/dgl/graphbolt/sampled_sub_graph.py (hunk 11 above)
Comment (peizhou001) [reply python/dgl/graphbolt/sampled_sub_graph.py]: Maybe `reverse` can intuitively reflect the mapping processing?

File: python/dgl/graphbolt/sampled_sub_graph.py (hunk 13 above)
Comment (peizhou001) [reply python/dgl/graphbolt/sampled_sub_graph.py]: Sure, added one example

File: python/dgl/graphbolt/__init__.py [hunk 17]
```diff
@@ -45,4 +45,4 @@ def load_graphbolt():
 
 load_graphbolt()
 
-SampledSubgraph = torch.classes.graphbolt.SampledSubgraph
+C_SampledSubgraph = torch.classes.graphbolt.SampledSubgraph
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/__init__.py:48]: Yes, it will be removed from here.

File: python/dgl/graphbolt/sampled_subgraph.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        the mapped ids of the column.
+        - If `reverse_column_node_ids` is a tensor: It represents the
+            original node ids.
+        - If `reverse_column_node_ids` is a dictionary: The keys should be
+            node type and the values should be corresponding original
+            heterogeneous node ids.
+        If present, it means column IDs are compacted, and `node_pairs`
+        column IDs match these compacted ones.
+        """
+        raise NotImplementedError
```
Comment (frozenbugs) [inline python/dgl/graphbolt/sampled_subgraph.py:42]: nit: Since the following 3 methods are optional, return None instead of raise error?

File: python/dgl/graphbolt/impl/csc_sampling_graph_sampled_subgraph.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+from dataclasses import dataclass
+from typing import Dict, Tuple, Union
+
+import torch
+
+from ..sampled_subgraph import SampledSubGraph
+
+
+@dataclass
+class CSCSamplingGraphSampledSubgraph(SampledSubGraph):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph_sampled_subgraph.py:12]: SampledSubgraphImpl

File: python/dgl/graphbolt/impl/csc_sampling_graph_sampled_subgraph.py
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph_sampled_subgraph.py:1]: sampled_subgraph_impl.py

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/impl/csc_sampling_graph_sampled_subgraph.py (hunk 19 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/csc_sampling_graph_sampled_subgraph.py:12]: make sense, changed for fix.

File: python/dgl/graphbolt/sampled_subgraph.py (hunk 18 above)
Comment (peizhou001) [reply python/dgl/graphbolt/sampled_subgraph.py:42]: Make sense, changed for fix.

File: python/dgl/graphbolt/sampled_subgraph.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
+"""Graphbolt sampled subgraph."""
+# pylint: disable= invalid-name
+from typing import Dict, Tuple
+
+import torch
+
+
+class SampledSubGraph:
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/sampled_subgraph.py:8]: Do we use `SampledSubgraph` or `SampledSubGraph`?
